experimental/devices/mfrc522/commands/low_level.go

```diff
@@ -0,0 +1,293 @@
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+package commands
+
+import (
+	"fmt"
+	"time"
+
+	"periph.io/x/periph/conn/gpio"
+	"periph.io/x/periph/conn/physic"
+	"periph.io/x/periph/conn/spi"
+)
+
+// AuthStatus indicates the authentication response, could be one of AuthOk,
+// AuthReadFailure or AuthFailure
+type AuthStatus byte
+
+// Card authentication status enum.
+const (
+	AuthOk AuthStatus = iota
+	AuthReadFailure
+	AuthFailure
+)
+
+// NewLowLevelSPI creates and initializes the RFID card reader attached to SPI.
+//
+// 	spiPort - the SPI device to use.
+// 	resetPin - reset GPIO pin.
+// 	irqPin - irq GPIO pin.
+func NewLowLevelSPI(spiPort spi.Port, resetPin gpio.PinOut, irqPin gpio.PinIn) (*LowLevel, error) {
+	if resetPin == nil {
+		return nil, wrapf("reset pin is not set")
+	}
+	if irqPin == nil {
+		return nil, wrapf("IRQ pin is not set")
+	}
+	spiDev, err := spiPort.Connect(10*physic.MegaHertz, spi.Mode0, 8)
+	if err != nil {
+		return nil, err
+	}
+	if err := resetPin.Out(gpio.High); err != nil {
+		return nil, err
+	}
+	if err := irqPin.In(gpio.PullUp, gpio.FallingEdge); err != nil {
+		return nil, err
+	}
+
+	dev := &LowLevel{
+		spiDev:   spiDev,
+		irqPin:   irqPin,
+		resetPin: resetPin,
+	}
+
+	return dev, nil
+}
+
+// LowLevel is a low-level handler of a MFRC522 RFID reader.
+type LowLevel struct {
+	resetPin gpio.PinOut
+	irqPin   gpio.PinIn
+	spiDev   spi.Conn
+}
+
+// String implements conn.Resource.
+func (r *LowLevel) String() string {
+	return fmt.Sprintf("Mifare MFRC522 [bus: %v, reset pin: %s, irq pin: %s]",
+		r.spiDev, r.resetPin.Name(), r.irqPin.Name())
+}
+
+// DevWrite sends data to a device.
+func (r *LowLevel) DevWrite(address int, data byte) error {
+	newData := []byte{(byte(address) << 1) & 0x7E, data}
+	return r.spiDev.Tx(newData, nil)
+}
+
+// DevRead gets data from a device.
+func (r *LowLevel) DevRead(address int) (byte, error) {
+	data := []byte{((byte(address) << 1) & 0x7E) | 0x80, 0}
+	out := make([]byte, len(data))
+	if err := r.spiDev.Tx(data, out); err != nil {
+		return 0, err
+	}
+	return out[1], nil
+}
+
+// CRC calculates the CRC of the data using the card chip.
+func (r *LowLevel) CRC(inData []byte) ([]byte, error) {
+	if err := r.ClearBitmask(DivIrqReg, 0x04); err != nil {
+		return nil, err
+	}
+	if err := r.SetBitmask(FIFOLevelReg, 0x80); err != nil {
+		return nil, err
+	}
+	for _, v := range inData {
+		if err := r.DevWrite(FIFODataReg, v); err != nil {
+			return nil, err
+		}
+	}
+	if err := r.DevWrite(CommandReg, PCD_CALCCRC); err != nil {
+		return nil, err
+	}
+	for i := byte(0xFF); i > 0; i-- {
+		n, err := r.DevRead(DivIrqReg)
+		if err != nil {
+			return nil, err
+		}
+		if n&0x04 > 0 {
+			break
+		}
+	}
+	lsb, err := r.DevRead(CRCResultRegL)
+	if err != nil {
+		return nil, err
+	}
+
+	msb, err := r.DevRead(CRCResultRegM)
+	if err != nil {
+		return nil, err
+	}
+	return []byte{lsb, msb}, nil
```

Comment:
I think returning a uint16 would be more sensible here. You could rename CRC to CRC16.